adapters/lockerdome/lockerdome.go

```diff
@@ -0,0 +1,137 @@
+package lockerdome
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const unexpectedStatusCodeMessage = "Unexpected status code: %d. Run with request.debug = 1 for more info"
+
+// Implements Bidder interface.
+type LockerDomeAdapter struct {
+	endpoint string
+}
+
+// MakeRequests makes the HTTP requests which should be made to fetch bids [from the bidder, in this case, LockerDome]
+func (adapter *LockerDomeAdapter) MakeRequests(openRTBRequest *openrtb.BidRequest, extraReqInfo *adapters.ExtraRequestInfo) (requestsToBidder []*adapters.RequestData, errs []error) {
+
+	numberOfImps := len(openRTBRequest.Imp)
+
+	if openRTBRequest.Imp == nil || numberOfImps == 0 { // lockerdometest/supplemental/empty_imps.json
+		err := &errortypes.BadInput{
+			Message: "No valid impressions in the bid request.",
+		}
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	for i := 0; i < numberOfImps; i++ {
+		// LockerDome currently only supports banner impressions, and requires data in the ext field.
+		if openRTBRequest.Imp[i].Banner == nil { // lockerdometest/supplemental/unsupported_imp_type.json
+			err := &errortypes.BadInput{
+				Message: "LockerDome does not currently support non-banner types.",
+			}
+			errs = append(errs, err)
+			return nil, errs
+		}
+		var bidderExt adapters.ExtImpBidder
+		err := json.Unmarshal(openRTBRequest.Imp[i].Ext, &bidderExt)
+		if err != nil { // lockerdometest/supplemental/no_ext.json
+			err = &errortypes.BadInput{
+				Message: "ext was not provided.",
+			}
+			errs = append(errs, err)
+			return nil, errs
+		}
+		var lockerdomeExt openrtb_ext.ExtImpLockerDome
+		err = json.Unmarshal(bidderExt.Bidder, &lockerdomeExt)
+		if err != nil { // lockerdometest/supplemental/no_adUnitId_param.json
+			err = &errortypes.BadInput{
+				Message: "ext.bidder.adUnitId was not provided.",
+			}
+			errs = append(errs, err)
+			return nil, errs
+		}
+		if lockerdomeExt.AdUnitId == "" { // lockerdometest/supplemental/empty_adUnitId_param.json
+			err := &errortypes.BadInput{
+				Message: "ext.bidder.adUnitId is empty.",
+			}
+			errs = append(errs, err)
+			return nil, errs
+		}
+	}
```

Comment:
Looks pretty good @margsliu, I like your clean and readable validation. One more suggestion and this is entirely optional: the current version of your code will discard the entire request even if only one element in its `Imp[]` array contains only one "bad" element, do we want that in our context? Some other adapters filter out the bad ones and post the request with only valid `Imp`s like:
```
request.Imp[]  { GOOD, BAD, GOOD, GOOD }
validImps[]    { GOOD, GOOD, GOOD }
```
and then they rewrite  `request.Imp[]` with `validImps[]` 
```
request.Imp[]  { GOOD, GOOD, GOOD }
```
Maybe something like:
```
   21 // MakeRequests makes the HTTP requests which should be made to fetch bids [from the bidder, in this case, LockerDome]
   22 func (adapter *LockerDomeAdapter) MakeRequests(openRTBRequest *openrtb.BidRequest, extraReqInfo *adapters.ExtraRequestInfo) (requestsToBidder []*adapters.RequestData, errs []error) {
   23
   24     numberOfImps := len(openRTBRequest.Imp)
   25
   26     if openRTBRequest.Imp == nil || numberOfImps == 0 { // lockerdometest/supplemental/empty_imps.json
   27         err := &errortypes.BadInput{
   28             Message: "No valid impressions in the bid request.",
   29         }
   30         errs = append(errs, err)
   31         return nil, errs
   32     }
   33
        + var goodImpPositions []int = make([]int)
   34     for i := 0; i < numberOfImps; i++ {
   35         // LockerDome currently only supports banner impressions, and requires data in the ext field.
   36         if openRTBRequest.Imp[i].Banner == nil { // lockerdometest/supplemental/unsupported_imp_type.json
   37             err := &errortypes.BadInput{
   38                 Message: "LockerDome does not currently support non-banner types.",
   39             }
   40             errs = append(errs, err)
   41   -         return nil, errs
        +         goodImpPositions = append(goodImpPositions,i)
   42         }
   43         var bidderExt adapters.ExtImpBidder
   44         err := json.Unmarshal(openRTBRequest.Imp[i].Ext, &bidderExt)
   45         if err != nil { // lockerdometest/supplemental/no_ext.json
   46             err = &errortypes.BadInput{
   47                 Message: "ext was not provided.",
   48             }
   49             errs = append(errs, err)
   50   -         return nil, errs
        +         goodImpPositions = append(goodImpPositions,i)
   51         }
   52         var lockerdomeExt openrtb_ext.ExtImpLockerDome
   53         err = json.Unmarshal(bidderExt.Bidder, &lockerdomeExt)
   54         if err != nil { // lockerdometest/supplemental/no_adUnitId_param.json
   55             err = &errortypes.BadInput{
   56                 Message: "ext.bidder.adUnitId was not provided.",
   57             }
   58             errs = append(errs, err)
   59   -         return nil, errs
        +         goodImpPositions = append(goodImpPositions,i)
   60         }
   61         if lockerdomeExt.AdUnitId == "" { // lockerdometest/supplemental/empty_adUnitId_param.json
   62             err := &errortypes.BadInput{
   63                 Message: "ext.bidder.adUnitId is empty.",
   64             }
   65             errs = append(errs, err)
   66   -         return nil, errs
        +         goodImpPositions = append(goodImpPositions,i)
   67         }
   68     }
   69
        + var validImps []Imp = make([]Imp)
        + for _, j := range goodImpPositions {
        +     validImps = append(validImps, openRTBRequest.Imp[j])
        + }
        + if len(validImps)==0 { 
        +     //More than fair if we decide to return error here, since no valid Imps were found
        + } else {
        +     //Substitute openRTBRequest.Imp with validImps
        + }
   70     openRTBRequestJSON, err := json.Marshal(openRTBRequest)
   71     if err != nil {
   72         errs = append(errs, err)
   73         return nil, errs
   74     }
   75
+  76 +--  8 lines: headers := http.Header{}---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
   84         Body:    openRTBRequestJSON,
   85         Headers: headers,
   86     }
   87
   88     requestsToBidder = append(requestsToBidder, requestToBidder)
   89
   90     return requestsToBidder, nil
   91
   92 }
```

Reply:
Thanks, I've pushed up changes that filter for valid Imps and have also adjusted the test cases accordingly. We are at 97% test coverage now.